### PR TITLE
feat(kuma-cp): extend policy matching API to work with egress and external services

### PR DESCRIPTION
**Pull request by lobkovilya**
Build rules for external services on zone egress.

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/plugins/interfaces.go [hunk 1]
```diff
@@ -95,3 +95,8 @@ type PolicyPlugin interface {
 	// You can access matching policies by using `proxy.Policies.Dynamic`.
 	Apply(rs *core_xds.ResourceSet, ctx xds_context.Context, proxy *core_xds.Proxy) error
 }
+
+type EgressPolicyPlugin interface {
+	PolicyPlugin
+	EgressMatchedPolicies(*core_mesh.ExternalServiceResource, xds_context.Resources) (core_xds.TypedMatchingPolicies, error)
```
Comment (lahabana) [inline pkg/core/plugins/interfaces.go:101]: Worth adding comments here please

File: pkg/plugins/policies/matchers/egress.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+			if !externalServiceSelectedByTargetRef(item.GetTargetRef(), es) {
+				continue
+			}
+			// convert 'to' policyItem to 'from' policyItem
+			artificial := &artificialPolicyItem{
+				conf:      item.GetDefault(),
+				targetRef: policy.GetSpec().(core_xds.Policy).GetTargetRef(),
+			}
+			toList = append(toList, core_xds.BuildPolicyItemsWithMeta([]core_xds.PolicyItem{artificial},
+				policy.GetMeta())...)
```
Comment (lahabana) [inline pkg/plugins/policies/matchers/egress.go:150]: Is this going to make things weird with the inspect API?

**Review by lahabana — COMMENTED** [review]
Look intuitively right

File: pkg/plugins/policies/matchers/egress.go (hunk 2 above)
Comment (lobkovilya) [reply pkg/plugins/policies/matchers/egress.go:150]: Not really, pieces of configs correspond to the actual existing policies. So they're correct, but maybe we should be careful with GUI

**Review by Automaat — APPROVED** [review]

**Pull request merged**